pkg/deviceshifu/deviceshifulwm2m/deviceshifulwm2m.go

```diff
@@ -0,0 +1,244 @@
+		return nil, fmt.Errorf("DeviceShifuLwM2M's namespace can't be empty")
+	}
+
+	base, mux, err := deviceshifubase.New(deviceShifuMetadata)
+	if err != nil {
+		return nil, err
+	}
+
+	lwM2MSetting := base.EdgeDevice.Spec.ProtocolSettings.LwM2MSetting
+	logger.Info("LwM2M endpoint is: %s", lwM2MSetting.EndpointName)
+	server, err := lwm2m.NewServer(*lwM2MSetting)
+	if err != nil {
+		return nil, err
+	}
+
+	ds := &DeviceShifuLwM2M{base: base, server: server}
+
+	go func() {
+		if err := server.Run(); err != nil {
+			logger.Fatalf("Error starting LwM2M server: %v", err)
+		}
+	}()
+
+	ds.instructionSettings = base.DeviceShifuConfig.Instructions.InstructionSettings
+	if ds.instructionSettings == nil {
+		ds.instructionSettings = &deviceshifubase.DeviceShifuInstructionSettings{}
+	}
+
+	if ds.instructionSettings.DefaultTimeoutSeconds == nil {
+		var defaultTimeoutSeconds = deviceshifubase.DeviceDefaultGlobalTimeoutInSeconds
+		ds.instructionSettings.DefaultTimeoutSeconds = &defaultTimeoutSeconds
+	}
+
+	lwM2MInstructions := CreateLwM2MInstructions(&base.DeviceShifuConfig.Instructions)
+
+	if deviceShifuMetadata.KubeConfigPath != deviceshifubase.DeviceKubeconfigDoNotLoadStr {
+		// switch for different Shifu Protocols
+		switch protocol := *base.EdgeDevice.Spec.Protocol; protocol {
+		case v1alpha1.ProtocolLwM2M:
+			for instruction, properties := range lwM2MInstructions.Instructions {
+				HandlerMetaData := &HandlerMetaData{
+					base.EdgeDevice.Spec,
+					instruction,
+					properties,
+				}
+				handler := DeviceCommandHandlerLwM2M{server, HandlerMetaData}
+				mux.HandleFunc("/"+instruction, handler.commandHandleFunc())
+
+				if properties.EnableObserve {
+					server.OnRegister(func() error {
+						return server.Observe(properties.ObjectId, func(data interface{}) {
+							logger.Infof("Observe data: %v", data)
+							// TODO need to push data to telemetry service
+						})
+					})
+				}
+			}
+		default:
+			logger.Errorf("EdgeDevice protocol %v not supported in deviceShifu LwM2M", protocol)
+			return nil, errors.New("wrong protocol not supported in deviceShifu LwM2M")
+		}
+	}
+	deviceshifubase.BindDefaultHandler(mux)
+
+	ds.base.UpdateEdgeDeviceResourcePhase(v1alpha1.EdgeDevicePending)
+	return ds, nil
+}
+
+// DeviceCommandHandlerLwM2M handler for http
+type DeviceCommandHandlerLwM2M struct {
+	server          *lwm2m.Server
+	HandlerMetaData *HandlerMetaData
+}
+
+// commandHandleFunc handle http request
+func (handler DeviceCommandHandlerLwM2M) commandHandleFunc() http.HandlerFunc {
+	handlerProperties := handler.HandlerMetaData.properties
+	handlerInstruction := handler.HandlerMetaData.instruction
+	handlerServer := handler.server
+
+	objectId := handlerProperties.ObjectId
+	return func(w http.ResponseWriter, r *http.Request) {
+		var respString string
+
+		switch r.Method {
+		case http.MethodPut:
+			requestBody, err := io.ReadAll(r.Body)
+			if err != nil {
+				http.Error(w, "Error on parsing body", http.StatusBadRequest)
```

Comment:
make the log lwm2m specific?
eg: " Error on parsing body for lwm2m deviceshifu." so do the following log.

Reply:
I think don't specific lwm2m in log, since this package only for it